chapters/cls-nonlinear.qmd

```diff
@@ -41,11 +53,885 @@ load("../RData/forested_data.RData")
+
+$$
+\begin{align}
+\boldsymbol{m}_i &= \phi\, \boldsymbol{m}_{i-1} + (1 - \phi)\, g(\boldsymbol{\theta}_i) \notag \\
+\boldsymbol{v}_i &= \rho\, \boldsymbol{v}_{i-1} + (1 - \rho)\, g(\boldsymbol{\theta}_i)^2 \notag 
+\end{align}
+$$
+
+Adam also tries to normalize these quantities so that the first few updates have larger values:   
+
+$$
+\begin{align}
+\boldsymbol{m}^*_i &= \frac{\boldsymbol{m}_i}{1 - \phi^2}\notag \\
+\boldsymbol{v}^*_i &= \frac{\boldsymbol{v}_i}{1 - \rho^2} \notag 
+\end{align}
+$$
+
+This helps offset the initialization of $\boldsymbol{m}_0 = \boldsymbol{v}_0 = \boldsymbol{0}$. At update $i$, Adam's proposal is
+
+$$
+\boldsymbol{\theta}_{i+1} = \boldsymbol{\theta}_i - \frac{\alpha}{\sqrt{\boldsymbol{v}^*_i + \epsilon}}\,\boldsymbol{m}^*_i.
+$$
+
+The results above show that Adam's path towards the optimal parameters is slightly wobbly compared to RMSprop but otherwise effective. 
+
+Both $\phi$ and $\rho$ are tunable in these models and typically range within [0.8, 0.99]. The constant learning rate $\alpha$ is also tunable. In @fig-sgd, $\phi=\rho=0.9$. AdaGrad, Adam, and RMSprop used $\alpha = 0.1$. For RMSprop, decreasing the rate to $\alpha = 0.08$ removed the oscillation effects shown in @fig-sgd.
+
+As previously alluded to, @loshchilov2017decoupled recognized that, for Adam, penalizing the loss function does not align with the definition of weight decay. Without a remedy, they found that Adam is not as effective as it could be. Their method, called AdamW, includes the penalty in the gradient definition. The gradient used to compute $\boldsymbol{m}_i$ and $\boldsymbol{v}_i$ is a combination of the previous "raw" gradient and the penalty ($\lambda$): 
+
+$$
+g(\boldsymbol{\theta}_i) = \nabla \psi(\boldsymbol{\theta}_{i}) - \lambda \boldsymbol{\theta}_{i}
+$$
+
+The updating equation also uses an extra term containing the penalty: 
+
+$$
+\boldsymbol{\theta}_{i+1} = \boldsymbol{\theta}_i - \alpha_i\left[\frac{\boldsymbol{m}^*_i}{\sqrt{\boldsymbol{v}^*_i + \epsilon}} + \lambda \boldsymbol{\theta}_{i}\right].
+$$
+
+Note that the learning rate $\alpha$ now has a subscript. They also recommend trying a cyclic learning rate scheduler with AdamW.  
+
+#### Other Methods {.unnumbered} 
+
+As previously mentioned, there are numerous other variations of SGD that will not be discussed here, including AdaMax [@kinga2015method], Nadam [@dozat2016], MadGrad [@defazio2021], and others.  
+
+### Newton's Method {#sec-newtons-method}
+
+As previously shown, Newton’s method (a.k.a., the Newton–Raphson method) uses both gradients and the Hessian matrix to update parameters. This approach can be very effective when the local loss surface can be approximated by a second-degree polynomial. When the local search includes a stationary point, the signs of the eigenvalues of the Hessian can characterize it: all non-negative imply a maximum, all non-positive imply a minimum, and mixed signs indicate a saddle point. 
```

Comment:
Originally, I was getting the best results from L-BFGS. However, I eventually figured out that `torch::optim_lbfgs()$step()` does _all_ of the steps instead of stopping at the end of an iteration. It turns out that, for `I` iterations, I was running `I` complete L-BFGS runs, each with 20 iterations. 

Not sure what to do about this in brulee just yet.